docs/csharp/tutorials/virtual-interface-methods.md

```diff
@@ -0,0 +1,121 @@
+
+An `ExtraFancyLight` might support both blink and timer functions directly:
+
+[!code-csharp[Override the blink and timer function](~/samples/csharp/tutorials/virtual-interface-methods/ExtraFancyLight.cs?name=SnippetExtraFancyLight)]
+
+The `HalogenLight` you created earlier does not support blinking. You just don't add the `IBlinkingLight` to the list of its supported interfaces.
+
+## Detect the light types using pattern matching
+
+Next, let's write some test code. You can make use of C#'s [pattern matching](../pattern-matching.md) feature to determine a light's capabilities by examining which interfaces it supports.  The following method exercises the supported capabilities of each light:
+
+[!code-csharp[Test a light's capabilities](~/samples/csharp/tutorials/virtual-interface-methods/Program.cs?name=SnippetTestLightFunctions)]
+
+The following code in your `Main` method creates each light type in sequence and tests that light:
+
+[!code-csharp[Test a light's capabilities](~/samples/csharp/tutorials/virtual-interface-methods/Program.cs?name=SnippetMainMethod)]
+
+## How the compiler determines best implementation
+
+This scenario shows a base interface without any implementations. Adding a method into the `ILight` interface introduces new complexities. The language rules governing default interface methods minimize the effect on the concrete classes that implement multiple derived interfaces. Let's enhance the original interface with a new method to show how that changes its use. Every indicator light can report its power status as an enumerated value:
+
+[!code-csharp[Enumeration for power status](~/samples/csharp/tutorials/virtual-interface-methods/ILight.cs?name=SnippetPowerStatus)]
+
+The default implementation assumes AC power:
+
+[!code-csharp[Report a default power status](~/samples/csharp/tutorials/virtual-interface-methods/ILight.cs?name=SnippetILightInterface)]
+
+These changes compile cleanly, even though the `ExtraFancyLight` declares support for the `ILight` interface and both derived interfaces, `ITimerLight` and `IBlinkingLight`. There's only one "closest" implementation declared in the `ILight` interface. Any class that declared an override would become the one "closest" implementation. You saw examples in the preceding classes that overrode the members of other derived interfaces.
+
+Avoid overriding the same method in multiple derived interfaces. Doing so creates an ambiguous method call whenever a class implements both derived interfaces. For example, if both the `IBlinkingLight` and `ITimerLight` implemented an override of `PowerStatus`, the `OverheadLight` would need to provide a more specific override. Otherwise, the compiler can't pick between the implementations in the two derived interfaces. You can usually avoid this situation by keeping interface definitions small and focused on one feature. In this scenario, each capability of a light is its own interface, multiple interfaces are only inherited by classes.
```

Comment:
```suggestion
Avoid overriding the same method in multiple derived interfaces. Doing so creates an ambiguous method call whenever a class implements both derived interfaces. For example, if both the `IBlinkingLight` and `ITimerLight` implemented an override of `PowerStatus`, the `OverheadLight` would need to provide a more specific override. Otherwise, the compiler can't pick between the implementations in the two derived interfaces. You can usually avoid this situation by keeping interface definitions small and focused on one feature. In this scenario, each capability of a light is its own interface; multiple interfaces are only inherited by classes.
```